programs/ssl/ssl_client2.c

```diff
@@ -982,6 +995,20 @@ int main( int argc, char *argv[] )
             if( opt.debug_level < 0 || opt.debug_level > 65535 )
                 goto usage;
         }
+#if defined(MBEDTLS_VERSION_C)
```

Comment:
I'm not sure I like having this do nothing if `MBEDTLS_VERSION_C` is not defined, as this will basically cause an error (printing usage and quitting) in that case.

Would it not be better to leave the option in, and make it print something like "version: unknown" in this case? 

I know this is is how it is done for other features, but then it could be expected that the user would know not to script usage of a feature which had not been compiled in. Checking for a version number I feel is a bit more generic. Others may disagree.

This obviously applies to all the `MBEDTLS_VERSION_C` guards added.

Reply:
The `mbedtls_version_xxx` functions are just a way to have a single copy of the version string in a build. The version string is always available in the source API via `version.h` through `MBEDTLS_VERSION_STRING` `MBEDTLS_VERSION_STRING_FULL`. In these test programs, we don't care about a few more bytes of code so we can just print out `MBEDTLS_VERSION_STRING_FULL` and not care if this leads to duplicating the string constant. (Which it probably doesn't, because most compilers unify string literals.)

Reply:
Thanks for the good suggestions. I'm not sure if I fully understand your improvements, just a few points to confirm. 

- The program doesn't have to depended on `MBEDTLS_VERSION_C`, as we can directly print out `MBEDTLS_VERSION_STRING_FULL` defined `mbedtls_config.h`. Moreover, build version is a generic feature not like other guards. So we can remove all the `MBEDTLS_VERSION_C` guards added in this change.
- Since `MBEDTLS_VERSION_STRING_FULL` is always available in `mbedtls_config.h`. It's not necessary to consider the case like version not defined. 

The final implementation would be
```
        else if( strcmp( p, "build_version" ) == 0 )
        {
            if( strcmp( q, "1" ) == 0 )
            {
                mbedtls_printf( "build version: %s (build %d)\n",
                                MBEDTLS_VERSION_STRING_FULL, MBEDTLS_VERSION_NUMBER);
                goto exit;
            }
        }
``` 
Alternately, we can remove digital build version and do this
```
        else if( strcmp( p, "build_version" ) == 0 )
        {
            if( strcmp( q, "1" ) == 0 )
            {
                mbedtls_printf( "build version: %s\n",
                                MBEDTLS_VERSION_STRING_FULL);
                goto exit;
            }
        }
```
Does this look good to you? Which one is more preferable?

Reply:
A new patch has been updated, please take a look.